lib/psql.js

```diff
@@ -24,31 +25,123 @@ function env (db) {
+  return {
+    username: 'bastion',
+    host: db.bastionHost,
+    privateKey: db.bastionKey,
+    dstHost: db.host,
+    dstPort: db.port,
+    localHost: localHost,
+    localPort: localPort
+  }
+}
+
+function buildTunnelPromise (db, dbTunnelConfig, timeout) {
+  return new Promise((resolve, reject) => {
+    // if necessary to tunnel, setup a tunnel
+    // see also https://github.com/heroku/heroku/blob/master/lib/heroku/helpers/heroku_postgresql.rb#L53-L80
+    if (db.bastionKey) {
+      tunnel(dbTunnelConfig, (err, tnl) => {
+        if (err) {
+          debug(err)
+          cli.error(`Unable to establish a secure tunnel to your database.`)
+          process.exit(1)
+        }
+        debug('Tunnel created')
+        resolve(tnl)
+      })
+    } else {
+      resolve()
+    }
+    setTimeout(() => reject('Establishing a secure tunnel timed out'), timeout)
+  }).catch((error) => {
+    cli.error(error)
+  })
+}
+
+function buildPsqlPromise (query, dbEnv, timeout) {
+  const stripEOF = require('strip-eof')
+  const {spawn} = require('child_process')
+  return new Promise((resolve, reject) => {
+    let psql = spawn('psql', ['-c', query], {env: dbEnv, encoding: 'utf8'})
+    psql.stdout.on('data', (data) => {
+      resolve(stripEOF(data))
+    })
+    psql.stderr.on('data', (data) => {
+      debug(stripEOF(data.toString('utf8')))
+      cli.error(`Unable to execute the command.`)
+      process.exit(1)
+    })
+    psql.on('error', (err) => {
+      debug(err)
+      handleError(err)
+    })
+    setTimeout(() => reject('psql call timed out'), timeout)
+  }).catch((error) => {
+    cli.error(error)
+  })
+}
+
+function buildPsqlInteractivePromise (dbEnv, prompt) {
+  const {spawn} = require('child_process')
+  return new Promise((resolve, reject) => {
+    let psql = spawn('psql',
+                     ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`],
+                     {env: dbEnv, stdio: 'inherit'})
+    psql.on('error', (err) => {
+      debug(err)
+      handleError(err)
+    })
+  })
+}
+
+function getConfigs (db) {
+  let dbEnv = env(db)
+  const dbTunnelConfig = tunnelConfig(db)
+  if (db.bastionKey) {
+    dbEnv = Object.assign(dbEnv, {
+      PGPORT: dbTunnelConfig.localPort,
+      PGHOST: dbTunnelConfig.localHost
+    })
+  }
+  return {
+    dbEnv: dbEnv,
+    dbTunnelConfig: dbTunnelConfig
+  }
+}
+
 function handleSignals () {
   process.once('SIGINT', () => {})
 }
 
-function * exec (db, query) {
-  const stripEOF = require('strip-eof')
-  const {spawnSync} = require('child_process')
+function * exec (db, query, timeout = 20000) {
   handleSignals()
   debug(query)
-  let {stdout, error: err, status} = spawnSync('psql', ['--command', query], {env: env(db), encoding: 'utf8', stdio: [0, 'pipe', 2]})
-  handleError(err)
-  if (status !== 0) process.exit(status)
-  return stripEOF(stdout)
+  let configs = getConfigs(db)
+  const tunnelPromise = buildTunnelPromise(db, configs.dbTunnelConfig, timeout)
```

Comment:
This section could be more direct:

```js
yield buildTunnelPromise(db, configs.dbTunnelConfig, timeout)
return yield buildPsqlPromise(query, configs.dbEnv, timeout)
```

Also the names of these functions should probably be something more like `tunnel()` or `createTunnel()` and `execPSQL()` or `psql()`. The fact it's generating a promise isn't as important as what the function does.